elastic/kibana
# Enable TSDB downsampling ILM configuration

## Summary

Close https://github.com/elastic/kibana/issues/130437

Enables TSDB downsampling action in ILM configuration. 
Actions can be set in the hot, warm, cold phases. 
 - It is hidden in warm and cold phases when searchable snapshots are enabled in hot phase (just like shrink, readonly, force merge actions). 

![Screenshot 2022-08-31 at 17 00 43](https://user-images.githubusercontent.com/7784120/187711538-e67999ff-5e25-4680-8b86-5ddcc64cca23.png)


Users need to specify an interval: 

![Screenshot 2022-08-31 at 17 00 48](https://user-images.githubusercontent.com/7784120/187711593-9540df96-a8d4-43cb-8621-4f50121ff414.png)

Intervals have the following validation: 
-  they must be greater than any previous interval and a multiple of that interval.
   - For example, on the warm phase, the user includes a 3-hour interval; if they choose to add a rollup action on the cold phase, the interval must be a multiple of 3h. It cannot be 5h, for example.
   - Another example: a user can have a rollup on the hot phase of 2d and then no rollup action on the warm phase, and then a rollup action on the cold phase, but it must be a multiple of 2d, for example, it cannot be 1d nor can it be 3d.

![Screenshot 2022-08-31 at 17 07 36](https://user-images.githubusercontent.com/7784120/187712990-03378b93-44af-409f-be1b-cb07f4cbd98b.png)


### How to test

(What I tested)

1. Create an ILM policy with Hot phase only. Enabled rollover, disable "use recommended settings" and set "maximum documents to 1 
2. Enable new "Downsample" action, set interval to 1 hour.
3. Save policy 
4. Then create a data stream template using Dev tools: https://github.com/csoulios/downsampling-demo/blob/main/downsample-ilm.es#L26-L169. Currently, it isn't possible to create this template using UI because of https://github.com/elastic/kibana/issues/139922.  Make sure you are attaching your ilm policy to the template. You can also do it through UI
7. Create data stream by injesting data using Dev tools: https://github.com/csoulios/downsampling-demo/blob/1e66a13bccc7d70aae7763bcc9c4499ec8b3fd02/downsample-ilm.es#L174-L194 . You will have to change timestamp in the data to the current hour 
8. Inspect indices and data streams. You should see them changed during the next hour. Inspect data of the downsampled stream in Discover


## Note

- Rollup of rollups is not yet implemented on es side. We should test everything end to end when it is ready. We wouldn't need to change ILM screen when it ready so we can merge without waiting for that.
- Ideally we should have a es docs link about ILM and downsampling to point to

### Release Notes

Enables time series downsampling action in ILM configuration
